src/hotspot/share/opto/cfgnode.cpp

```diff
@@ -2376,8 +2406,70 @@ Node *PhiNode::Ideal(PhaseGVN *phase, bool can_reshape) {
       }
     }
 
-    // This restriction is temporarily necessary to ensure termination:
-    if (!saw_self && adr_type() == TypePtr::BOTTOM)  merge_width = 0;
+    // There are cases with circular dependencies between bottom Phis
+    // and MergeMems. Below is a minimal example.
+    //
+    //               +------------+
+    //               |            |
+    // (base_memory) v            |
+    //              MergeMem      |
+    //                 |          |
+    //                 v          |
+    //                Phi (this)  |
+    //                 |          |
+    //                 v          |
+    //                Phi         |
+    //                 |          |
+    //                 +----------+
+    //
+    // Here, we cannot break the circularity through saw_self = true as there
+    // are two Phis involved. Repeatedly splitting the Phis through the
+    // MergeMem leads to non-termination. We check for non-termination below.
+    bool terminates = true;
+    // Only check for non-termination if necessary.
+    if (!saw_self && adr_type() == TypePtr::BOTTOM &&
+        merge_width > Compile::AliasIdxRaw) {
+      ResourceMark rm;
+      VectorSet visited;
+      Node_List worklist;
+      worklist.push(this);
+      visited.set(this->_idx);
+      auto add_to_worklist = [&](Node* input) {
```

Comment:
Suggestion: rename to `maybe_add_to_worklist` to stress that it is a conditional addition.